### PR TITLE
 multi-client: failover enhancements

### PR DESCRIPTION
**Pull request by iurii-ssv**
This PR is meant to improve multi-client failover capabilities, previous implementation is using a simple heuristic of "client being synced" as the best way to predict if the client can serve API request or not (aka "active" client) + deactivating client under certain failure scenarios during API request execution, the issue is - that behavior doesn't cover some of the cases we've encountered in practice (eg. fetch-duties call was just timing out due to ~12s deadline we set, but at the same time we weren't trying to switch to another client presumably because the one we were trying reported it's synced).

Client being **synced** doesn't guarantee it's gonna serve API requests faithfully in a stable manner, hence a better heuristic is needed. A simple scoring system implemented in this PR (as seamlessly as possible on top of the previous sync-based approach to keep the code well-aligned with upstream) would do a better job at differentiating between good & bad clients, now failover works roughly like this:
- every client is classified into two categories: active/inactive based on its sync status (checked in background periodically)
- only currently active clients can be chosen to process API requests
- all clients (active or deactivated) have a score assigned to them, that score reflects how well they are expected to serve API requests based on the past history, there is no differentiation between various API endpoints (eg. failing `endpoint1` call has the same effect on client score as failing `endpoint2` call)
- the client with the highest score always serves API requests

Besides the added scoring mechanism this PR is also addressing the following edge-cases:
- whenever a client request times out with `context.DeadlineExceeded` error the previous implementation will simply keep trying that very same client every time without any failover, this is not ideal because this can very well happen regardless of what timeout(deadline) value the caller chooses due to variety of reasons: this particular client being overloaded at the moment, temporary network issue or crash while sync status hasn't updated yet, etc. - there is no reason not to try another client in that scenario
- (could potentially modify this behavior) additionally, whenever a client returns `4**` type error response the previous implementation will simply keep trying that very same client every time without any failover - while the request itself is the most likely problem in this case, there is also a chance a bug in the client is the root cause for the request failing - so I'm not 100% we want to "trust" single client response (in SSV setting it would probably be better to fail over to another client just in case it's a client-bug, but I'm not changing this behavior for now)

I think this change should also remove the need for implementing things like https://github.com/ssvlabs/ssv/pull/2339 (@nkryuchkov do you think we should partially-revert stuff added in https://github.com/ssvlabs/ssv/pull/2339 ?).

Note, this PR is **complementary** to https://github.com/attestantio/go-eth2-client/pull/192 but instead of "trying to make sync reporting more reliable" it targets the failover algorithm itself to make it more generic & robust.

### CODE REVIEW

**Comment by iurii-ssv** [comment]
@greptileai could you review this PR for me pls ?

File: multi/client.go [hunk 1]
```diff
@@ -161,25 +180,36 @@ func (s *Service) doCall(ctx context.Context, call callFunc, errHandler errHandl
 		return nil, errors.New("no clients to which to make call")
 	}
 
+	// Sort in desc order (clients with the highest scores come first), client-scores
+	// might change concurrently to sorting - but it's best to use the latest data anyway.
+	slices.SortFunc(activeClients, func(a, b consensusclient.Service) int {
+		aScore := s.scoreClient(a.Address())
```
[the remark this answers is not in the document]
Comment (iurii-ssv) [reply multi/client.go:183]: Yeah, that's what I'm referring to in the comment (changing the score while sorting is kinda fine),

but the lock-contention indeed might be better addressed by locking once

File: multi/client.go [hunk 2]
```diff
@@ -130,6 +132,16 @@ func (s *Service) activateClient(ctx context.Context, client consensusclient.Ser
 	s.setConnectionsMetric(ctx, len(s.activeClients), len(s.inactiveClients))
 }
 
+// scoreClient returns client score.
+func (s *Service) scoreClient(clientAddr string) int {
+	return s.clientScores[clientAddr]
+}
+
+// penalizeClient client score.
```
Comment (nkryuchkov) [inline multi/client.go:140]:
```suggestion
// penalizeClient penalizes client score.
```

```suggestion
// penalizeClient decreases client score.
```

File: multi/client.go [hunk 3]
```diff
@@ -161,25 +173,24 @@ func (s *Service) doCall(ctx context.Context, call callFunc, errHandler errHandl
...
+	activeClients = s.activeClientsSortedByScore()
+
 	var err error
 	var res any
 	for _, client := range activeClients {
 		log := log.With().Str("client", client.Name()).Str("address", client.Address()).Logger()
 		res, err = call(ctx, client)
 		if err != nil {
-			log.Trace().Err(err).Msg("Potentially deactivating client due to error")
+			log.Trace().Err(err).Msg(fmt.Sprintf("Potentially failing over from client %s due to error", client.Address()))
```
Comment (nkryuchkov) [inline multi/client.go:185]:
```suggestion
			log.Trace().Err(err).Msgf("Potentially failing over from client %s due to error", client.Address())
```

File: multi/client.go [hunk 4]
```diff
@@ -161,25 +173,24 @@ func (s *Service) doCall(ctx context.Context, call callFunc, errHandler errHandl
...
 			var apiErr *api.Error
 			switch {
 			case errors.As(err, &apiErr) && statusCodeFamily(apiErr.StatusCode) == 4:
-				log.Trace().Err(err).Msg("Not deactivating client on user error")
+				log.Trace().Err(err).Msg(fmt.Sprintf("Not failing over from client %s on user error", client.Address()))
```
Comment (nkryuchkov) [inline multi/client.go:189]: ditto

File: multi/client.go [hunk 5]
```diff
@@ -189,8 +219,8 @@ func (s *Service) doCall(ctx context.Context, call callFunc, errHandler errHandl
 				failover, err = errHandler(ctx, client, err)
 			}
 			if failover {
-				log.Debug().Err(err).Msg("Deactivating client on error")
-				s.deactivateClient(ctx, client)
+				log.Debug().Err(err).Msg(fmt.Sprintf("Failing over from client %s on error", client.Address()))
+				s.penalizeClient(client.Address())
```
[the remark this answers is not in the document]
Comment (nkryuchkov) [reply multi/client.go:206]: I think the case could happen quite often, e.g., if CL's EL is down, so CL reports as working but doesn't. As I remember, I caught this case when I was testing the initial multi-client version. But I agree the current version looks good enough, we just need to test it intensively to make sure we won't get any performance degradation in some of the edge cases

**Review by nkryuchkov — DISMISSED** [review]
good job, let's make sure this is tested well enough with real nodes

**Review by nkryuchkov — APPROVED** [review]

**Review by y0sher — APPROVED** [review]
